Tell docker which images to use as a cache when building

Bidski opened the pull request.

TrentHouliston commented on tools/dockerise/__init__.py:
```diff
@@ -40,20 +40,26 @@ def build_platform(platform):
             "build",
             ".",
+            "--cache-from",
+            "{},{}".format(local_tag, remote_tag),
```
```suggestion
            ','.join([local_tag, remote_tag]),
```

Bidski replied:
Thats why `local_tag` is specified first so it will take its cache from that image (assuming that `local_tag` is your locally modified image and `remote_tag` is master)

TrentHouliston commented on tools/dockerise/__init__.py:
```diff
@@ -88,24 +106,32 @@ def get_selected_platform():
+            if len(names) == 0:
+                cprint("ERROR the currently selected platform is a dangling tag.", "orange", attrs=["bold"])
+                cprint(
+                    "      The system is unable to work out what platform this was and will need to be reset orange",
```
Will need to be reset orange! How scary!

TrentHouliston reviewed (dismissed):
Presuming this works, LGTM

Bidski commented:
This should be ready to merge, but I think we need to upgrade docker on the buildkite machines to 19.03.14 or later. @JosephusPaye @ysims @KipHamiltons

KipHamiltons commented on tools/dockerise/__init__.py:
```diff
@@ -106,24 +122,33 @@ def get_selected_platform():
+            names = [
+                tag.split(":")[-1]
+                for tag in img_info[0]["RepoTags"]
+                if tag != "{}:selected".format(repository) and tag.startswith("{}:".format(repository))
```
I think this is a little more readable
```suggestion
                if tag != f"{repository}:selected" and tag.startswith(f"{repository}:")
```

KipHamiltons commented on tools/dockerise/__init__.py:
```diff
@@ -175,17 +206,16 @@ def run(rebuild, clean, platform, network, **kwargs):
                             subprocess.call(
-                                ["docker", "image", "tag", tag, "{}:selected".format(repository)],
-                                stdout=subprocess.DEVNULL,
+                                ["docker", "tag", tag, "{}:selected".format(repository)], stdout=subprocess.DEVNULL,
```
```suggestion
                                ["docker", "tag", tag, f"{repository}:selected"], stdout=subprocess.DEVNULL,
```

Bidski replied:
I feel like this is a change that should be made in a separate PR in all python files to maintain consistency

KipHamiltons replied:
My opinion is that it's only more readable in some situations though

KipHamiltons reviewed (approved):
I like `f"{id}"` format strings, so I have added a couple of suggestions with those :P 

LGTM. I removed all of my images and ran `./b target generic` and it pulled without issue :smile:

Bidski commented:
> I like `f"{id}"` format strings, so I have added a couple of suggestions with those :P
> 
> LGTM. I removed all of my images and ran `./b target generic` and it pulled without issue

Pulling is not what this resolves, but building. Try running these commands
```sh
docker container prune -af
docker image prune -af
docker builder prune -af
./b target generic
```

If all works properly then when you get to the build phase of `./b target generic` then every layer should be cached and building should be over pretty quickly. (you need to remove the builder and image caches to ensure you dont get false positives)

KipHamiltons commented:
Ahh right. I used the wrong word there. I had removed the containers and images. I'll try again with the builds pruned too

KipHamiltons commented:
So when I try this on my (dual core macbook) laptop, it _seems_ to hang here: 
```
$ ./b target generic
generic: Pulling from nubots/nubots
Digest: sha256:9df2eccf442b48e5443f58f96555fdc2deeb671fb6ac5441750cbab99b6e4a73
Status: Image is up to date for nubots/nubots:generic
docker.io/nubots/nubots:generic
```
Then when I press ctrl + c once, it starts building, and it _seems_ to hang either here:
```
[+] Building 4.0s (92/92) FINISHED                                              
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 37B                                        0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 34B                                           0.0s
 => [internal] load metadata for docker.io/library/archlinux:base-2021011  3.6s
 => importing cache manifest from nubots/nubots:generic                    0.0s
 => importing cache manifest from nubots:generic                           0.0s
 => [internal] load build context                                          0.0s
 => => transferring context: 1.91kB                                        0.0s
 => [ 1/84] FROM docker.io/library/archlinux:base-20210110.0.13332@sha256  0.0s
 => CACHED [ 2/84] RUN echo "Server=https://archive.archlinux.org/repos/2  0.0s
 => CACHED [ 3/84] RUN pacman -Syyuu --noconfirm --needed --overwrite *    0.0s
 => CACHED [ 4/84] COPY usr/local/bin/install-package /usr/local/bin/inst  0.0s
 => CACHED [ 5/84] RUN install-package     wget     sudo     python     p  0.0s
 => CACHED [ 6/84] RUN echo $(python -c "import site; print(site.getsitep  0.0s
 => CACHED [ 7/84] COPY etc/xdg/pip/pip.conf /etc/xdg/pip/pip.conf         0.0s
 => CACHED [ 8/84] COPY etc/ld.so.conf.d/usrlocal.conf /etc/ld.so.conf.d/  0.0s
 => CACHED [ 9/84] RUN ldconfig                                            0.0s
 => CACHED [10/84] RUN cd /usr/local && ln -sf lib lib64                   0.0s
 => CACHED [11/84] COPY usr/local/toolchain/generate_toolchains.py /usr/l  0.0s
 => CACHED [12/84] COPY usr/local/toolchain/generate_generic_toolchain.py  0.0s
 => CACHED [13/84] RUN python /usr/local/generate_toolchain.py --prefix /  0.0s
 => CACHED [14/84] COPY usr/local/bin/download-and-extract /usr/local/bin  0.0s
 => CACHED [15/84] COPY usr/local/bin/install-from-source /usr/local/bin/  0.0s
 => CACHED [16/84] RUN ln -s /usr/local/bin/install-from-source /usr/loca  0.0s
 => CACHED [17/84] COPY usr/local/toolchain/generate_generic_toolchain.py  0.0s
 => CACHED [18/84] RUN python /usr/local/generate_toolchain.py --prefix /  0.0s
 => CACHED [19/84] COPY usr/local/toolchain/generate_generic_toolchain.py  0.0s
 => CACHED [20/84] RUN python /usr/local/generate_toolchain.py --prefix /  0.0s
 => CACHED [21/84] RUN install-from-source https://www.zlib.net/zlib-1.2.  0.0s
 => CACHED [22/84] RUN BUILD_FILE_DIR='llvm' install-cmake-from-source ht  0.0s
 => CACHED [23/84] RUN install-cmake-from-source https://github.com/Khron  0.0s
 => CACHED [24/84] COPY usr/local/package/opencl-clang.sh /usr/local/pack  0.0s
 => CACHED [25/84] RUN /usr/local/package/opencl-clang.sh https://github.  0.0s
 => CACHED [26/84] RUN install-from-source https://ftp.gnu.org/pub/gnu/nc  0.0s
 => CACHED [27/84] COPY usr/local/package/intel-graphics-compiler.sh /usr  0.0s
 => CACHED [28/84] RUN /usr/local/package/intel-graphics-compiler.sh http  0.0s
 => CACHED [29/84] RUN install-cmake-from-source https://github.com/intel  0.0s
 => CACHED [30/84] RUN install-from-source https://xorg.freedesktop.org/r  0.0s
 => CACHED [31/84] RUN install-meson-from-source https://dri.freedesktop.  0.0s
 => CACHED [32/84] RUN install-from-source https://github.com/intel/libva  0.0s
 => CACHED [33/84] RUN install-cmake-from-source https://github.com/intel  0.0s
 => CACHED [34/84] COPY --chown=root:root etc/OpenCL/vendors/intel.icd /e  0.0s
 => CACHED [35/84] COPY usr/local/package/intel-compute-runtime.sh /usr/l  0.0s
 => CACHED [36/84] RUN /usr/local/package/intel-compute-runtime.sh https:  0.0s
 => CACHED [37/84] COPY usr/local/package/opencl-headers.sh /usr/local/pa  0.0s
 => CACHED [38/84] RUN /usr/local/package/opencl-headers.sh https://githu  0.0s
 => CACHED [39/84] RUN install-from-source https://github.com/KhronosGrou  0.0s
 => CACHED [40/84] RUN install-package ruby                                0.0s
 => CACHED [41/84] RUN install-from-source https://github.com/OCL-dev/ocl  0.0s
 => CACHED [42/84] RUN install-package gcc-fortran                         0.0s
 => CACHED [43/84] COPY usr/local/package/openblas/generic.sh /usr/local/  0.0s
 => CACHED [44/84] RUN /usr/local/package/openblas.sh https://github.com/  0.0s
 => CACHED [45/84] RUN install-cmake-from-source https://downloads.source  0.0s
 => CACHED [46/84] COPY usr/local/include/armadillo_bits/config.hpp /usr/  0.0s
 => CACHED [47/84] COPY usr/local/package/eigen3/deprecated-copy.patch /u  0.0s
 => CACHED [48/84] COPY usr/local/package/eigen3/cxx17-overalign.patch /u  0.0s
 => CACHED [49/84] RUN install-from-source-with-patches https://gitlab.co  0.0s
 => CACHED [50/84] RUN install-from-source     https://github.com/gperfto  0.0s
 => CACHED [51/84] RUN install-package protobuf                            0.0s
 => CACHED [52/84] RUN install-autotools-from-source https://github.com/p  0.0s
 => CACHED [53/84] RUN install-package yasm                                0.0s
 => CACHED [54/84] RUN install-from-source https://github.com/libjpeg-tur  0.0s
 => CACHED [55/84] RUN install-from-source https://github.com/jbeder/yaml  0.0s
 => CACHED [56/84] RUN  install-from-source https://github.com/fmtlib/fmt  0.0s
 => CACHED [57/84] RUN install-header-from-source https://github.com/catc  0.0s
 => CACHED [58/84] RUN install-from-source https://github.com/libusb/libu  0.0s
 => CACHED [59/84] RUN install-from-source http://xmlsoft.org/sources/lib  0.0s
 => CACHED [60/84] RUN install-from-source https://github.com/libffi/libf  0.0s
 => CACHED [61/84] RUN install-from-source https://gitlab.gnome.org/GNOME  0.0s
 => CACHED [62/84] RUN install-meson-from-source https://github.com/Aravi  0.0s
 => CACHED [63/84] RUN install-cmake-from-source https://github.com/libuv  0.0s
 => CACHED [64/84] RUN install-from-source https://github.com/Fastcode/NU  0.0s
 => CACHED [65/84] RUN install-from-source https://github.com/ianlancetay  0.0s
 => CACHED [66/84] RUN install-from-source ftp://ftp.alsa-project.org/pub  0.0s
 => CACHED [67/84] RUN install-from-source https://github.com/espeak-ng/p  0.0s
 => CACHED [68/84] COPY usr/local/package/espeak.sh /usr/local/package/es  0.0s
 => CACHED [69/84] RUN /usr/local/package/espeak.sh https://github.com/es  0.0s
 => CACHED [70/84] RUN pip install     pylint     termcolor     stringcas  0.0s
 => CACHED [71/84] RUN install-package     arm-none-eabi-gcc     arm-none  0.0s
 => CACHED [72/84] RUN pip install --upgrade     cmake-format==0.6.10      0.0s
 => CACHED [73/84] COPY etc/sudoers.d/user /etc/sudoers.d/user             0.0s
 => CACHED [74/84] RUN chmod 440 /etc/sudoers.d/user                       0.0s
 => CACHED [75/84] RUN groupadd --gid 501 nubots     && useradd --uid 501  0.0s
 => CACHED [76/84] RUN install -d -m 0755 -o nubots -g nubots /home/nubot  0.0s
 => CACHED [77/84] COPY --chown=nubots:nubots home/nubots/.ssh/id_rsa /ho  0.0s
 => CACHED [78/84] COPY --chown=nubots:nubots home/nubots/.ssh/id_rsa.pub  0.0s
 => CACHED [79/84] COPY --chown=nubots:nubots home/nubots/.ssh/config /ho  0.0s
 => CACHED [80/84] COPY --chown=nubots:nubots home/nubots/.gdbinit /home/  0.0s
 => CACHED [81/84] RUN chmod 600 /home/nubots/.ssh/id_rsa /home/nubots/.s  0.0s
 => CACHED [82/84] RUN install -d -m 0755 /home/nubots/NUbots              0.0s
 => CACHED [83/84] RUN install -d -m 0777 /home/nubots/build               0.0s
 => CACHED [84/84] WORKDIR /home/nubots/NUbots                             0.0s
 => exporting to image                                                     0.1s
 => => exporting layers                                                    0.0s
 => => writing image sha256:6c7b050498ef19ea60bfbe58ec29efce1348c2c61f27b  0.0s
 => => naming to docker.io/library/nubots:generic                          0.0s
 => exporting cache                                                        0.0s
 => => preparing build cache for export   
```
or just a little later:
```
.
.
 => => naming to docker.io/library/nubots:generic                          0.0s
 => exporting cache                                                        0.0s
 => => preparing build cache for export                                    0.0s
------
 > importing cache manifest from nubots:generic:
------
```
I waited for 20 minutes and nothing happened. htop had my cpu usage at basically idle levels. My net speed is fine.
Do you have any ideas?

Bidski commented:
I have seen this before, just wait and it will probably finish (eventually). I am not happy with this though, do you have any IO activity?

KipHamiltons commented:
atm I am sitting on `importing cache manifest` and I have less than 1MB/s of IO when I am not doing anything else
